app/grandchallenge/evaluation/models.py

```diff
@@ -863,25 +912,38 @@ def assign_permissions(self):
         assign_perm("view_evaluation", admins_group, self)
         assign_perm("change_evaluation", admins_group, self)
 
-        if self.submission.phase.challenge.hidden:
+        if self.submission.phase.hidden:
+            viewer_group = []
+            non_viewer_group = [
+                self.submission.phase.challenge.participants_group,
+                Group.objects.get(
+                    name=settings.REGISTERED_AND_ANON_USERS_GROUP_NAME
+                ),
+            ]
+        elif self.submission.phase.challenge.hidden:
             viewer_group = self.submission.phase.challenge.participants_group
-            non_viewer_group = Group.objects.get(
-                name=settings.REGISTERED_AND_ANON_USERS_GROUP_NAME
-            )
+            non_viewer_group = [
+                Group.objects.get(
+                    name=settings.REGISTERED_AND_ANON_USERS_GROUP_NAME
+                )
+            ]
         else:
             viewer_group = Group.objects.get(
                 name=settings.REGISTERED_AND_ANON_USERS_GROUP_NAME
             )
-            non_viewer_group = (
+            non_viewer_group = [
                 self.submission.phase.challenge.participants_group
-            )
+            ]
 
         if self.published:
-            assign_perm("view_evaluation", viewer_group, self)
+            if viewer_group:
+                assign_perm("view_evaluation", viewer_group, self)
         else:
-            remove_perm("view_evaluation", viewer_group, self)
+            if viewer_group:
+                remove_perm("view_evaluation", viewer_group, self)
 
-        remove_perm("view_evaluation", non_viewer_group, self)
+        for non_viewers in non_viewer_group:
+            remove_perm("view_evaluation", non_viewers, self)
```

Comment:
This is quite difficult to reason about now. Could you simplify it?

Something like:
```
all_users_can_view = self.published and not self.submission.phase.hidden and not self.submission.phase.challenge.hidden

if all_users_can_view:
  assign_perm("view_evaluation", all_user_group, self)
else:
  remove_perm("view_evaluation", all_user_group, self)

participants_can_view = self.published and not self.submission.phase.hidden

if participants_can_view:
  assign_perm("view_evaluation", participants_group, self)
else:
  remove_perm("view_evaluation", participants_group, self)
```